internal/olm/client.go

```diff
@@ -0,0 +1,467 @@
+	crdResources, err := c.getCRDs(ctx, version)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch CRDs: %s", err)
+	}
+
+	log.Infof("Fetching resources for version %s", version)
+	olmResources, err := c.getOLM(ctx, version)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch resources: %s", err)
+	}
+
+	resources := append(crdResources, olmResources...)
+	return resources, nil
+}
+
+func (c Client) getStatus(ctx context.Context, objs []unstructured.Unstructured) Status {
+	var rss []ResourceStatus
+	for _, obj := range objs {
+		nn := types.NamespacedName{
+			Namespace: obj.GetNamespace(),
+			Name:      obj.GetName(),
+		}
+		u := unstructured.Unstructured{}
+		u.SetGroupVersionKind(obj.GroupVersionKind())
+		err := c.KubeClient.Get(ctx, nn, &u)
+		rs := ResourceStatus{
+			NamespacedName: nn,
+			GVK:            obj.GroupVersionKind(),
+		}
+		if err != nil {
+			rs.Error = err
+		} else {
+			rs.Resource = &u
+		}
+		rss = append(rss, rs)
+	}
+
+	return Status{Resources: rss}
+}
+
+func (c Client) resolveVersion(ctx context.Context, version string) (string, error) {
+	var url string
+	if version == "latest" {
+		url = fmt.Sprintf("%s/%s", c.BaseReleasesAPIURL, version)
+	} else {
+		url = fmt.Sprintf("%s/tags/%s", c.BaseReleasesAPIURL, version)
+	}
+
+	resp, err := c.doRequest(ctx, url)
+	if err != nil {
+		return version, fmt.Errorf("request failed: %s", err)
+	}
+	defer resp.Body.Close()
+
+	data, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return version, fmt.Errorf("read response body: %s", err)
+	}
+
+	var d struct {
+		TagName string `json:"tag_name"`
+	}
+	if err := json.Unmarshal([]byte(data), &d); err != nil {
+		return version, fmt.Errorf("unmarshal json: %s", err)
+	}
+	return d.TagName, nil
+}
+
+func (c Client) getCRDs(ctx context.Context, version string) ([]unstructured.Unstructured, error) {
+	resp, err := c.doRequest(ctx, c.crdsURL(version))
+	if err != nil {
+		return nil, fmt.Errorf("request failed: %s", err)
+	}
+	defer resp.Body.Close()
+	return decodeResources(resp.Body)
+}
+
+func (c Client) getOLM(ctx context.Context, version string) ([]unstructured.Unstructured, error) {
+	resp, err := c.doRequest(ctx, c.olmURL(version))
+	if err != nil {
+		return nil, fmt.Errorf("request failed: %s", err)
+	}
+	defer resp.Body.Close()
+	return decodeResources(resp.Body)
+}
+
+func (c Client) doRequest(ctx context.Context, url string) (*http.Response, error) {
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return nil, fmt.Errorf("create request: %s", err)
+	}
+	req = req.WithContext(ctx)
+	resp, err := c.HttpClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed GET '%s': %s", url, err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("failed GET '%s': unexpected status code %d, expected %d", url, resp.StatusCode, http.StatusOK)
+	}
+	return resp, nil
+}
+
+func (c Client) getBaseURL(version string) string {
```

Comment:
Is this the base URL of the github release download?

Reply:
Yep, I'll update it to `getBaseDownloadURL` so it's more clear.